README
Ham-Fldigi version 0.01
=======================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2012 by Andy Smith

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Ham::Fldigi.

0.01  Mon Feb  6 17:18:42 2012
	- original version; created by h2xs 1.23 with options
		-X Ham::Fldigi



ham-fldigi.pl
#!/usr/bin/perl

#==============================================================================
# ham-fldigi.pl
# v0.001
# (c) 2012 Andy Smith, M0VKG
#==============================================================================
# DESCRIPTION
# Example usage of Ham::Fldigi.
#==============================================================================

use Ham::Fldigi;
use Data::Dumper;
use Carp;

# Create a new Ham::Fldigi object, and then use it to create a new
# Ham::Fldigi::Client object.
my $f = new Ham::Fldigi('LogLevel' => 4, 'LogFile' => './debug.log', 'LogPrint' => 1, 'LogWrite' => 1);
my $c = $f->client('Hostname' => "localhost", 'Port' => 7362, 'Name' => "test");

# If $c is undef, there's been a problem...
if(!defined($c)) { croak("Couldn't connect to fldigi!"); };

# Set the modem to BPSK125 and transmit some text.
#$c->command("modem.set_by_name", "BPSK125");
#$c->send("CQ CQ CQ DE M0QQQ M0QQQ M0QQQ KN");
#my $s = $f->shell('Client' => $c);
my $s = $f->shell();
#$s->client($c);
$s->start;


lib/Ham/Fldigi.pm
#!/usr/bin/perl

#==============================================================================
# Ham::Fldigi
# v0.001
# (c) 2012 Andy Smith, M0VKG
#==============================================================================
# DESCRIPTION
# Perl extensions for managing Fldigi instances
#==============================================================================
# SYNOPSIS
# use Ham::Fldigi;
# my $client = Ham::Fldigi::Client->new('localhost', 7362, 'default');
# $client->modem("BPSK125");
# $client->send("CQ CQ CQ DE M0VKG M0VKG M0VKG KN");
#==============================================================================

# Perl documentation is provided inline in pod format.
# To view, run:-
# perldoc Ham::Fldigi

=head1 NAME

Ham::Fldigi - Perl extensions for managing Fldigi instances

=head1 SYNOPSIS

	use Ham::Fldigi;

=head1 DESCRIPTION

This module itself doesn't do much - see C<Ham::Fldigi::Client> for details.

=head2 EXPORT

None by default.
=cut

package Ham::Fldigi;

use 5.012004;
use strict;
use warnings;

our $VERSION = '0.001';

use Moose;
use Ham::Fldigi::Client;
use Ham::Fldigi::Shell;
use POE qw(Wheel::Run);
use base qw(Ham::Fldigi::Debug);

has 'clients' => (is => 'ro', isa => 'HashRef[Ham::Fldigi::Client');

=head1 CONSTRUCTORS

=head2 Fldigi->new([I<LogLevel> => n, ] [I<LogFile> => filename, ] [I<LogPrint> => (0|1), ] [I<LogWrite> => (0|1)]) 

Creates a new B<Ham::Fldigi> object with the specified options.

=item * 

I<LogLevel> is an integer between 0 and 4, with 0 being no logging at all, 1 for errors, 2 for warnings, 3 for notices and 4 for debugging. This defaults to B<2>, which will display and log errors and warnings.

=item *

I<LogFile> is the path to the logfile that will be written to.

=item *

I<LogPrint> is whether to print log messages to screen or not.

=item *

I<LogWrite> is whether to log messages to the logfile or not.

=cut

sub new {
	
	# Get the class name
	my($class) = shift;
	my(%params) = @_;

	my $self = {
		'version' => $VERSION,
	};

	if(defined($params{'LogLevel'})) {
		$Ham::Fldigi::Debug::debug_level = $params{'LogLevel'};
	}
	if(defined($params{'LogFile'})) {
		$Ham::Fldigi::Debug::debug_file = $params{'LogFile'};
	}
	if(defined($params{'LogPrint'})) {
		$Ham::Fldigi::Debug::debug_print = $params{'LogPrint'};
	}
	if(defined($params{'LogWrite'})) {
		$Ham::Fldigi::Debug::debug_write = $params{'LogWrite'};
	}

	bless $self, $class;

	$self->debug("Constructor called. Version is ".$VERSION.".");
	$self->debug("Returning...");
	return $self;
}

=head1 MEHTODS

=head2 Fldigi->client(I<hostname>, I<port>, I<name>)

Creates a new B<Ham::Fldigi::Client> object with the specified arguments. See C<Ham::Fldigi::Client> for more details.

=cut

sub client {

	my ($self, %params) = @_;

	my $c = Ham::Fldigi::Client->new(%params);
	$self->{clients}{$c->name} = $c;

	return $c;
}

sub shell {

	my ($self, %params) = @_;

	$params{'Parent'} = $self;

	my $s = Ham::Fldigi::Shell->new(%params);

	return $s;

}

1;
__END__

=head1 SEE ALSO

The source code for this module is hosted on Github at L<https://github.com/m0vkg/Perl-Ham-Fldigi>.

=head1 AUTHOR

Andy Smith M0VKG, E<lt>andy@m0vkg.org.ukE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by Andy Smith

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/Ham/Fldigi/Debug.pm
#==============================================================================
# Ham::Fldigi
# v0.001
# (c) 2012 Andy Smith, M0VKG
#==============================================================================
# DESCRIPTION
# Colourful debug messages for Ham::Fldigi
#==============================================================================
# SYNOPSIS
# use Ham::Fldigi;
# my $client = Ham::Fldigi::Client->new('localhost', 7362, 'default');
# $client->modem("BPSK125");
# $client->send("CQ CQ CQ DE M0VKG M0VKG M0VKG KN");
#==============================================================================

# Perl documentation is provided inline in pod format.
# To view, run:-
# perldoc Ham::Fldigi::Debug

=head1 NAME

Ham::Fldigi::Debug - Colourful debug messages for Ham::Fldigi

=head1 SYNOPSIS

  use Ham::Fldigi;
	use base qw(Ham::Fldigi::Debug)

	...
	$self->debug("This is a debug message!");

=head1 DESCRIPTION

Colourful debug/error messages for Ham::Fldigi.

By default, none of the calls output anything - either to screen or to a logfile. This is intentional so as not to interfere when Ham::Fldigi is used by third-party scripts.

If you do want to make use of the logging functions, see the documentation for C<Ham::Fldigi>'s B<new()>.

=head2 EXPORT

B<debug>, B<error>, B<warning> and B<notice>.

=cut

package Ham::Fldigi::Debug;

use 5.012004;
use strict;
use warnings;

use Term::ANSIColor;
use Data::Dumper;
use DateTime;

our @ISA = qw(Exporter);
our %EXPORT_TAGS = ( 'all' => [ qw() ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw(debug error warning notice);

our $debug_level = 0;
our $debug_print = 0;
our $debug_file = "/dev/null";
our $debug_write = 0;

# _timestamp
# Create timestamps for our log messages
sub _timestamp {

	my ($self) = @_;

	my $dt = DateTime->now;

	my $timestamp = $dt->ymd." ".$dt->hms;

	return $timestamp;
}

=head1 METHODS

=head2 debug(I<message>)

Print a debug message.

=cut

sub debug {
  
  my ($self, $msg) = @_;

	my $class;
  if(ref($self)) {
		$class = ref $self;
  } else {
		$class = $self;
  }

	my @caller = split('::', (caller(1))[3]);
	my $sub = $caller[$#caller];

	my $logmsg = "[".$self->_timestamp."] ".(color 'blue')."+++".(color 'reset')." ".(color 'magenta').$class.(color 'reset')." (".(color 'green').$sub.color('reset')."): ".$msg."\n";

	if($debug_level ge 4) {
		if($debug_print eq 1) {
			print STDERR "\r".$logmsg;
		}
		if($debug_write eq 1) {
			open (LOGFILE, '>>'.$debug_file);
			print LOGFILE $logmsg;
		}
	}
}

=head2 error(I<message>)

Print an error message.

=cut

sub error {
  
  my ($self, $msg) = @_;

	my $class;
  if(ref($self)) {
		$class = ref $self;
  } else {
		$class = $self;
  }

	my @caller = split('::', (caller(1))[3]);
	my $sub = $caller[$#caller];

	my $logmsg = "[".$self->_timestamp."] ".(color 'red')."!!!".(color 'reset')." ".(color 'magenta').$class.(color 'reset')." (".(color 'green').$sub.color('reset')."): ".$msg."\n";

	if($debug_level ge 1) {
		if($debug_print eq 1) {
			print STDERR "\r".$logmsg;
		}
		if($debug_write eq 1) {
			open (LOGFILE, '>>'.$debug_file);
			print LOGFILE $logmsg;
		}
	}
}


=head2 warning(I<message>)

Print a warning message.

=cut

sub warning {
  
  my ($self, $msg) = @_;

	my $class;
  if(ref($self)) {
		$class = ref $self;
  } else {
		$class = $self;
  }

	my @caller = split('::', (caller(1))[3]);
	my $sub = $caller[$#caller];

	my $logmsg = "[".$self->_timestamp."] ".(color 'yellow')."xxx".(color 'reset')." ".(color 'magenta').$class.(color 'reset')." (".(color 'green').$sub.color('reset')."): ".$msg."\n";

	if($debug_level ge 2) {
		if($debug_print eq 1) {
			print STDERR "\r".$logmsg;
		}
		if($debug_write eq 1) {
			open (LOGFILE, '>>'.$debug_file);
			print LOGFILE $logmsg;
		}
	}
}


=head2 notice(I<message>)

Print a notice message.

=cut

sub notice {
  
  my ($self, $msg) = @_;

	my $class;
  if(ref($self)) {
		$class = ref $self;
  } else {
		$class = $self;
  }

	my @caller = split('::', (caller(1))[3]);
	my $sub = $caller[$#caller];

	my $logmsg = "[".$self->_timestamp."] ".(color 'green')."---".(color 'reset')." ".(color 'magenta').$class.(color 'reset')." (".(color 'green').$sub.color('reset')."): ".$msg."\n";

	if($debug_level ge 3) {
		if($debug_print eq 1) {
			print STDERR "\r".$logmsg;
		}
		if($debug_write eq 1) {
			open (LOGFILE, '>>'.$debug_file);
			print LOGFILE $logmsg;
		}
	}
}

1;
__END__

=head1 SEE ALSO

The source code for this module is hosted on Github at L<https://github.com/m0vkg/Perl-Ham-Fldigi>.

=head1 AUTHOR

Andy Smith M0VKG, E<lt>andy@m0vkg.org.ukE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by Andy Smith

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/Ham/Fldigi/Client.pm
#!/usr/bin/perl

#==============================================================================
# Ham::Fldigi::Client
# v0.001
# (c) 2012 Andy Smith, M0VKG
#==============================================================================
# DESCRIPTION
# This module handles communications with a running fldigi instance
#==============================================================================
# SYNOPSIS
# use Ham::Fldigi;
# my $client = Ham::Fldigi::Client->new('localhost', 7362, 'default');
# $client->modem("BPSK125");
# $client->send("CQ CQ CQ DE M0VKG M0VKG M0VKG KN");
#==============================================================================

# Perl documentation is provided inline in pod format.
# To view, run:-
# perldoc Client.pm

=head1 NAME

Ham::Fldigi::Client - Perl extension for communicating with Fldigi.

=head1 SYNOPSIS

  use Ham::Fldigi;
  my $f = new Ham::Fldigi;
	my $c = new Ham::Fldigi::Client('localhost', '7362', 'example');

	$c->command('fldigi.version');

=head1 DESCRIPTION

This module is for communicating with individual Fldigi instances.

It uses Fldigi's XMLRPC service, which usually runs on localhost:7362, providing support for it has been compiled in.

=head2 EXPORT

None by default.
=cut

package Ham::Fldigi::Client;

use 5.012004;
use strict;
use warnings;

use Moose;
use Data::GUID;
use RPC::XML::Client;
use Data::Dumper;
use Time::HiRes qw( usleep );
use POE qw( Wheel::Run );
use base qw(Ham::Fldigi::Debug);

has 'hostname' => (is => 'rw');
has 'port' => (is => 'rw');
has 'name' => (is => 'rw');
has 'url' => (is => 'ro');
has 'id' => (is => 'ro');
has '_xmlrpc' => (is => 'ro');
has '_session' => (is => 'ro');
has '_buffer_tx' => (is => 'ro');
has '_buffer_rx' => (is => 'ro');

our $VERSION = '0.001';

=head1 CONSTRUCTORS

=head2 Client->new(I<hostname>, I<port>, I<name>)

Creates a new B<Ham::Fldigi::Client> object with the specified arguments.

=cut

sub new {
	
	# Get our name, and set an ID
	my $class = shift;
	my $g = Data::GUID->new;

	# Fill in the class ID and version
	my $self =  {
		'version' => $VERSION,
		'id' => $g->as_string,
	};

	# Bless self
	bless $self, $class;

	$self->debug("Constructor called. Version ".$VERSION.", with ID ".$self->id.".");

	# Grab the passed client details
	my (%params) = @_;
	if(defined($params{'Hostname'})) {
		$self->hostname($params{'Hostname'});
	} else {
		$self->hostname("localhost");
	}
	if(defined($params{'Port'})) {
		$self->port($params{'Port'});
	} else {
		$self->port(7362);
	}
	if(defined($params{'Name'})) {
		$self->name($params{'Name'});
	} else {
		$self->name($self->id);
	}

	$self->{url} = 'http://'.$self->hostname.':'.$self->port.'/RPC2';
	$self->debug("Hostname is ".$self->hostname.", port is ".$self->port." and name is ".$self->name.".");
	$self->debug("XML-RPC URL is ".$self->url.".");

	# Initialise the RPC::XML::Client object
	$self->{_xmlrpc} = RPC::XML::Client->new($self->url);

	# Check connectivity with the fldigi client by checking the version
	$self->debug("Checking connectivity with fldigi client at ".$self->url."...");
	my $fldigi_version = $self->version;

	if(defined($fldigi_version)) {
		$self->debug("Version is ".$fldigi_version.".");
	} else {
		return undef;
	}

	$self->debug("Returning...");
	return $self;
}

=head1 METHODS

=head2 Client->command(I<command>, [I<arguments>])

Send command I<command>, with optional arguments I<arguments>.

=cut

sub command {

	my ($self, $cmd, $args) = @_;

	# If $args is unset, set it with an empty value
	if(!defined($args)) { $args = "" };

	$self->debug("Making XMLRPC call '".$cmd."' (args: ".$args.") to http://".$self->hostname.":".$self->port."/RPC2...");
	my $r = $self->_xmlrpc->simple_request($cmd, $args);

	# Check for undef response, which means there's been an error
	if($RPC::XML::ERROR ne "") {
		$self->error("Error talking to ".$self->url."!");
		$self->error("RPC::XML::Client reports: ".$RPC::XML::ERROR);
		return undef;
	}

	# If there's no response from XMLRPC, set it to '(null)'
	if(!defined($r)) { $r = "(null)"; };

	$self->debug("Response from XMLRPC request is: ".$r);
	return $r;
}

=head2 Client->version

Query for the version of Fldigi.

=cut

sub version {

	my ($self) = @_;
	my $r = $self->command('fldigi.version');

	return $r;

}

=head2 Client->send("I<text>")

Clears the TX window of any existing text, adds the text passed as I<text>, and then transmits. It automatically adds a '^r' at the end to tell Fldigi to switch back to RX afterwards.

=cut

sub send {

	my ($self, $text) = @_;

	# Clear the TX window of any existing text, add our text and then switch to TX.
	# We add a '^r' on the end to tell fldigi to stop once it's transmitted all
	# the waiting text.
	$self->command("text.clear_tx");
	$self->command("text.add_tx", $text."^r");
	$self->command("main.tx");

}

1;
__END__

=head1 SEE ALSO

The source code for this module is hosted on Github at L<https://github.com/m0vkg/Perl-Ham-Fldigi>.

=head1 AUTHOR

Andy Smith M0VKG, E<lt>andy@m0vkg.org.ukE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by Andy Smith

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=cut
